fix: Allow Team Event editing by the person who created it

hariombalhara opened the pull request:
## What does this PR do?
- Allow editing/deleting of a team event by the member who created it.
- Owner/Admin would still have the editing/deleting access for all team events, no matter who created it.

 **This is applicable to all newly created events only because team events are still not having `userId` set**

Fixes #2035

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?
- [x] Created a team event(Event1) and try to update/delete it. It should be allowed. Check that **Readonly** tag goes away in Event Types Listing
- [x] Login as Team owner, you should be able to update/delete the same event(Event1). Check that  **Readonly** tag isn't therein Event Types Listing
- [x] Login as any other member of the team, you should be able to see event Event1. It should have a  **Readonly** tag and you shouldn't be able to update/delete the event. Backend would reject the request.
- [x] Check that non-team event is also editable by the creator.

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

hariombalhara committed:
Allow Team Event editing by the person who created it

hariombalhara committed:
Deduplicate editing check

hariombalhara committed:
Add comments

hariombalhara commented on packages/prisma/migrations/20220305052238_adds_event_creator/migration.sql:
```diff
@@ -0,0 +1,47 @@
+ALTER TABLE "EventType" ADD CONSTRAINT "EventType_creatorId_fkey" FOREIGN KEY ("creatorId") REFERENCES "users"("id") ON DELETE SET NULL ON UPDATE CASCADE;
+
+-- RenameIndex
+ALTER INDEX "Booking.uid_unique" RENAME TO "Booking_uid_key";
```
Not sure why these indices renaming happened. This is automatically generated

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix/allow-self-event-editing

hariombalhara committed:
Remove unnecessary autogenerated queries

hariombalhara committed:
Merge branch 'main' into fix/allow-self-event-editing

hariombalhara committed:
Merge branch 'main' into fix/allow-self-event-editing

hariombalhara committed:
Debug failing test

hariombalhara committed:
Merge branch 'main' into fix/allow-self-event-editing

hariombalhara committed:
Debug failing test -Revert

Commit:
Merge branch 'main' into fix/allow-self-event-editing

pumfleet reviewed (commented):
Upon testing, this seems to work fine. However I did not test what happens when this migration is run on top of existing event types, which will not have a creator ID set. Does this prevent everyone except admins from editing it?

agustif commented on apps/web/next.config.js:
```diff
@@ -18,6 +18,7 @@ if (!process.env.NEXT_PUBLIC_APP_URL) {
   process.env.NEXT_PUBLIC_APP_URL = process.env.BASE_URL;
 }
 process.env.NEXT_PUBLIC_BASE_URL = process.env.BASE_URL;
+console.warn("VARS", process.env.PLAYWRIGHT_SECRET);
```
Maybe we could copy userId -> creatorId on the migration, not sure if it would bring some edge cases like events not created by current owner maybe?

agustif commented:
> Upon testing, this seems to work fine. However I did not test what happens when this migration is run on top of existing event types, which will not have a creator ID set. Does this prevent everyone except admins from editing it?

Maybe we could copy userId -> creatorId on the migration, not sure if it would bring some edge cases like events not created by current owner maybe?

The rest looks good!

agustif commented on apps/web/pages/api/auth/forgot-password.ts:
```diff
@@ -63,7 +63,7 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
 
     await sendPasswordResetEmail(passwordEmail);
-
+    console.log("FORGOT PASSWORD", process.env.PLAYWRIGHT_SECRET);
```
We could add some no-console eslint rules, I always have to ctrl-f and double check before committing lol

agustif reviewed (changes requested):
Any idea why the zod/webhook empty file is on the file changes?

<img width="954" alt="image" src="https://user-images.githubusercontent.com/6601142/158424161-357d635e-725d-4f70-a174-797fce8d303e.png">
I think @zomars fixed the zod auto-generated files being picked up.

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix/allow-self-event-editing

hariombalhara committed:
Revert accidental changes

hariombalhara committed:
Merge branch 'fix/allow-self-event-editing' of github.com:hariombalhara/cal.com into fix/allow-self-event-editing

hariombalhara commented:
> > Upon testing, this seems to work fine. However I did not test what happens when this migration is run on top of existing event types, which will not have a creator ID set. Does this prevent everyone except admins from editing it?
> 
> Maybe we could copy userId -> creatorId on the migration, not sure if it would bring some edge cases like events not created by current owner maybe?
> 
> The rest looks good!

As per my understanding userId is not set now. Only `users` is set and that actually has a list of attendees for the event type. So, there is no information in PG that can reliably tell us the creator of the event.

I have removed accidental code changes and this PR is ready for review again

zomars committed:
Merge branch 'main' into pr/2066

zomars committed:
Merge branch 'main' into pr/2066

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

agustif commented:
Hey @hariombalhara code looks good to me. 
Cant we gent the event type owner from userId which lives in db in user event type relation table?

From schema.prisma

relation("user_eventtype")
  userId                  Int?

agustif reviewed (commented):
LGTM, but I think userId should be available in eventType and might be === to new creatorId? Maybe we can migrate the data from there

hariombalhara commented:
@agustif regarding that userId thing, I created a new event type and checked DB studio and this is what I saw
Notice, the last one USERIDTEST. It doesn't have `userId` set. It has users though.

![image](https://user-images.githubusercontent.com/1780212/164454093-f1496ff2-216a-41a8-8a0c-7a6fd20dec86.png)

zomars commented:
@hariombalhara @agustif Currently we have a many-to-many relationship from users and event types. Hence why we don't have `userId` set. I think this is being used for team events? I could be wrong tho. Tagging @emrysal to see if he knows more about this data model.

hariombalhara commented:
@zomars This has been fixed recently in https://github.com/calcom/cal.com/pull/2599

Now I can see userId, I would change the code to use that instead of creatorId
<img width="1457" alt="Screenshot 2022-05-17 at 3 48 41 PM" src="https://user-images.githubusercontent.com/1780212/168789273-31076fe0-9950-4126-93cc-ba05d768bba3.png">

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix/allow-self-event-editing

hariombalhara committed:
Fix TS errors

hariombalhara committed:
Fix Ts errors and deletion failure message

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

hariombalhara commented:
Updated the PR as now userId is being set. It was missing for team event, fixed that as well. I am not sure why specific check for team event was added earlier.

zomars commented:
> Updated the PR as now userId is being set. It was missing for team event, fixed that as well. I am not sure why specific check for team event was added earlier.


Does this means that this fix won’t apply to already created records?

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into fix/allow-self-event-editing

hariombalhara committed:
Remove unused variable

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing

kodiakhq[bot] committed:
Merge branch 'main' into fix/allow-self-event-editing
